### PR TITLE
reworked unittests

### PR DESCRIPTION
**Pull request by RalphBariz**
https://github.com/dcarp/cmake-d/pull/23

there is still a small thing I don't get away. doesn't affect the functionality is just about the if you asked

### CODE REVIEW

**Comment by dcarp** [comment]
Unfortunately with this PR the tests don't run anymore :(. Please verify it with `cd tests && ./run_tests_linux.sh`.

File: cmake-d/UseDUnittest.cmake [hunk 1]
```diff
@@ -7,31 +7,64 @@
...
+    set(_TEST_TARGET "test_${_TARGET}")
+    
+    get_target_property(_TARGET_TEST_SOURCES ${_TARGET} SOURCES)
+    if(_TARGET_TEST_SOURCES)    
+        get_target_property(_TARGET_INCLUDE_DIRECTORIES ${_TARGET} INCLUDE_DIRECTORIES)
+        get_target_property(_TARGET_LINK_LIBRARIES ${_TARGET} LINK_LIBRARIES)
 
+        _add_unittests(${_TEST_TARGET} "${_TARGET_TEST_SOURCES}" "${_TARGET_INCLUDE_DIRECTORIES}" "${_TARGET_LINK_LIBRARIES}")
+    else()
+        message(SEND_ERROR "cannot add unittests for a source file less target")
```
Comment (dcarp) [inline cmake-d/UseDUnittest.cmake]: "Cannot add unittests for target ${_TARGET} with no source files"

Comment (RalphBariz) [reply cmake-d/UseDUnittest.cmake]: hmm... at my system they do
http://picpaste.com/Screenshot_from_2017-12-04_23-14-34.png

Comment (RalphBariz) [reply cmake-d/UseDUnittest.cmake]: I'm on cmake 3.10

Comment (dcarp) [reply cmake-d/UseDUnittest.cmake]: sorry, I was too laconic... I meant to suggest a "better" message text.

File: cmake-d/UseDUnittest.cmake [hunk 2]
```diff
@@ -7,31 +7,64 @@
...
 endmacro()
+
+macro(add_unittests _SOURCE)
+    set(_TEST_TARGET "test_${_SOURCE}")
+
+    get_directory_property(_INCLUDE_DIRECTORIES INCLUDE_DIRECTORIES)
+    get_directory_property(_LINK_LIBRARIES LINK_LIBRARIES)
+    
+    _add_unittests(${_TEST_TARGET} "${_SOURCE}" "${_INCLUDE_DIRECTORIES}" "${_LINK_LIBRARIES}")
+endmacro()
```
Comment (dcarp) [inline cmake-d/UseDUnittest.cmake]: new line at the end of file is missing

File: tests/app_9/CMakeLists.txt [hunk 3]
```diff
@@ -1,14 +1,20 @@
+cmake_minimum_required(VERSION 3.0)
+project (test D)
 include(UseDUnittest)
 
+set(CMAKE_BUILD_TYPE DEBUG)
```
Comment (dcarp) [inline tests/app_9/CMakeLists.txt]: `tests/run_tests_linux.sh` tests all the build types. Don't fix it here, just use the script (cmake_minimum_required(VERSION 3.0); project (test D); enable_testing() are also not needed)

File: tests/app_9/CMakeLists.txt [hunk 4]
```diff
@@ -1,14 +1,20 @@
...
+
+enable_testing()
+
 # TODO
 # how to rerun cmake if a new file is created
 # I don't want to miss new unittests
 
 # add all files
-file(GLOB files RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} "*.d")
+file(GLOB files RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} "module*.d")
```
Comment (dcarp) [inline tests/app_9/CMakeLists.txt]: Globing is [not recommended](https://cmake.org/cmake/help/v3.10/command/file.html) anymore, simply specify the two files.
```
set(FILES
  ${CMAKE_CURRENT_SOURCE_DIR} "moduleA.d
  ${CMAKE_CURRENT_SOURCE_DIR} "moduleB.d)
```

Comment (RalphBariz) [reply tests/app_9/CMakeLists.txt]: Just restored yours :-D

**Comment by RalphBariz** [comment]
Today I fear I wont. Tomorrow.

**Comment by RalphBariz** [comment]
updated, there should be a commit from today. ./run_tests_linux.sh is working for me

**Comment by dcarp** [comment]
LGTM, Thanks!

**Pull request merged**